followthemoney/graph.py

```diff
@@ -60,50 +72,59 @@ class Edge(object):
     __slots__ = ['id', 'weight', 'source_id', 'target_id',
                  'prop', 'proxy', 'schema', 'graph']
 
-    def __init__(self, graph, source, target, proxy=None, prop=None, value=None):  # noqa
+    def __init__(self, graph: 'Graph', source: Node, target: Node,
+                 proxy: Optional[EntityProxy]=None, prop: Optional[Property]=None,
+                 value: Optional[str]=None):  # noqa
         self.graph = graph
         self.id = f"{source.id}<>{target.id}"
         self.source_id = source.id
         self.target_id = target.id
         self.weight = 1.0
-        self.prop = prop
-        self.proxy = proxy
-        self.schema = None
+        self.prop: Optional[Property] = prop
+        self.proxy: Optional[EntityProxy] = proxy
+        self.schema: Optional[Schema] = None
         if prop is not None:
             self.weight = prop.specificity(value)
         if proxy is not None:
             self.id = f"{source.id}<{proxy.id}>{target.id}"
             self.schema = proxy.schema
 
     @property
-    def source(self):
+    def source(self) -> Optional[Node]:
         return self.graph.nodes.get(self.source_id)
 
     @property
-    def source_prop(self):
+    def source_prop(self) -> Optional[Property]:
```

Comment:
Edge must have a source prop, if you want we can check if it's a valid edge before.

Reply:
I should do the following then:
1. `self.schema.source_prop.reverse` cannot be `None` (given `self.schema.source_prop` exists)
2. `self.prop` cannot be `None`

Reply:
Okay, I missed a thing. In order to initialize an edge, we need to have at least a `prop` OR a `proxy` (to initialize the schema)